crypto/aes_gcm.go

```diff
@@ -0,0 +1,161 @@
+	Pbkdf2Iterations int    `json:"pbkdf2iterations,omitempty" toml:"pbkdf2iterations"`
+	Pbkdf2SaltLen    int    `json:"pbkdf2saltlen,omitempty" toml:"pbkdf2saltlen"`
+}
+
+type AESGCMStore struct {
+	Salt       []byte `json:"salt"`
+	Nonce      []byte `json:"nonce"`
+	Ciphertext []byte `json:"ciphertext"`
+}
+
+func DefaultAESGCMSettings() *AESGCMSettings {
+	return &AESGCMSettings{
+		KeyLen:           32,
+		Pbkdf2Hash:       "sha512",
+		Pbkdf2Iterations: 100000,
+		Pbkdf2SaltLen:    16,
+	}
+}
+
+func NewAESGCMClient(config Config, settings AESGCMSettings) (*AESGCMClient, error) {
+	return &AESGCMClient{
+		config:   config,
+		settings: settings,
+	}, nil
+}
+
+func (c *AESGCMClient) keyLen() int {
+	keyLen := c.settings.KeyLen
+	switch keyLen {
+	default:
+		if keyLen != 0 {
+			fmt.Println("Invalid keyLen, using default")
+		}
+		return 32
+	case 16:
+	case 24:
+	case 32:
+	}
+	return keyLen
+}
+
+func (c *AESGCMClient) pbkdf2HashFunc() func() hash.Hash {
+	pbkdf2Hash := c.settings.Pbkdf2Hash
+	switch pbkdf2Hash {
+	default:
+		if pbkdf2Hash != "" {
+			fmt.Println("Invalid PBKDF2 Hash, using default")
+		}
+		fallthrough
+	case "sha512":
+		return sha512.New
+	}
+}
+
+func (c *AESGCMClient) pbkdf2SaltLen() int {
+	return c.settings.Pbkdf2SaltLen
+}
+
+func (c *AESGCMClient) deriveKeyWithSalt(password []byte, salt []byte) []byte {
+	keyLen := c.keyLen()
+	pbkdf2Iterations := c.settings.Pbkdf2Iterations
+	return pbkdf2.Key([]byte(password), salt, pbkdf2Iterations, keyLen, c.pbkdf2HashFunc())
+}
+
+func (c *AESGCMClient) deriveKey(password []byte) ([]byte, []byte, error) {
+	salt := make([]byte, c.pbkdf2SaltLen())
+	if _, err := rand.Read(salt); err != nil {
+		return nil, nil, err
+	}
+
+	return c.deriveKeyWithSalt(password, salt), salt, nil
+}
+
+func (c *AESGCMClient) Decrypt(data []byte, password []byte) (plaintext []byte, err error) {
+	var store AESGCMStore
+	if err := json.Unmarshal(data, &store); err != nil {
+		return nil, err
+	}
+
+	key := c.deriveKeyWithSalt(password, store.Salt)
+
+	ac, err := aes.NewCipher(key)
+	if err != nil {
+		return
+	}
+
+	gcm, err := cipher.NewGCM(ac)
+	if err != nil {
+		return
+	}
+
+	plaintext, err = gcm.Open(nil, store.Nonce, store.Ciphertext, nil)
+	if err != nil {
+		return nil, &errors.SafeDecryptionFailed{}
+	}
+
+	return
+}
+
+func (c *AESGCMClient) Encrypt(plaintext []byte, password []byte) (data []byte, err error) {
+	key, salt, err := c.deriveKey(password)
+	if err != nil {
+		return
+	}
+
+	ac, err := aes.NewCipher(key)
+	if err != nil {
+		return
+	}
+
+	gcm, err := cipher.NewGCM(ac)
+	if err != nil {
+		return
+	}
+
+	nonce := make([]byte, gcm.NonceSize())
+	if _, err = rand.Read(nonce); err != nil {
+		return
+	}
+
+	ciphertext := gcm.Seal(nil, nonce, plaintext, nil)
+
+	store := AESGCMStore{
+		Salt:       salt,
+		Nonce:      nonce,
+		Ciphertext: ciphertext,
+	}
+	data, err = json.Marshal(store)
+	if err != nil {
+		return
+	}
+
+	return
```

Comment:
We might consider explicitly returning the values on `Encrypt` and `Decrypt`, rather than implicitly using the named return values. It's a little hard to read, especially with the error handling on 152. 

The named return values on the `crypto.Client` interface help clarify the intention, but I'm leaning against using named return values on the implementation- it may be harder to read and it's inconsistent with the rest of the code base.

FWIW, I think I'm the guilty one who did it first ;)